File: tests/fixtures/resolution/csharp/shared_method_names.cs
// ADR-011 §4 resolution fixture (C#): several classes share a method NAME, so a bare-name
// resolver (ADR-021) cannot settle the call — only the receiver's TYPE says which target.
// This is the exact case receiver-type inference exists for. Ground truth in the sibling
// .resolution.json is authored from the source semantics, never copied from resolver output.
namespace Shop
{
    public class OrderRepo
    {
        public void Save() { }   // shared name: OrderRepo also defines Save
        public void Load() { }   // UNIQUE name: only OrderRepo defines Load
    }

    public class UserRepo
    {
        public void Save() { }   // shared name: collides with OrderRepo.Save
    }

    public class Service
    {
        private OrderRepo _orders;

        // ── The lift: ambiguous name, disambiguated only by the receiver type ──
        public void SaveViaField()          { _orders.Save(); }               // → OrderRepo.Save
        public void SaveViaParam(UserRepo u){ u.Save(); }                     // → UserRepo.Save
        public void SaveViaNew()            { var o = new OrderRepo(); o.Save(); }  // → OrderRepo.Save

        // ── Control: a UNIQUE name resolves with or without the hint (no lift, no regression) ──
        public void LoadUnique()            { _orders.Load(); }               // → OrderRepo.Load

        // ── Prefer-unknown (§2): must stay unresolved, in BOTH regimes ──
        public void ExternalReceiver(System.Text.StringBuilder sb) { sb.Save(); }  // external type → null
        public void ChainedReceiver()       { GetOrders().Save(); }          // chained receiver → null

        public OrderRepo GetOrders() { return _orders; }
    }
}

File: tests/fixtures/resolution/cpp/shared_method_names.cpp
// ADR-011 §4 resolution fixture (C++): mirrors the C# case — several classes share a method
// NAME, so only the receiver's TYPE disambiguates the call. Receiver reached via -> and .
// Ground truth in the sibling .resolution.json is authored from the source semantics.
namespace shop {

class OrderRepo {
public:
    void Save();   // shared name
    void Load();   // UNIQUE name: only OrderRepo defines Load
};

class UserRepo {
public:
    void Save();   // shared name: collides with OrderRepo::Save
};

class Service {
    OrderRepo* orders_;
public:
    // ── The lift: ambiguous name, disambiguated only by the receiver type ──
    void saveViaField()            { orders_->Save(); }        // → OrderRepo::Save
    void saveViaParam(UserRepo* u) { u->Save(); }              // → UserRepo::Save
    void saveViaLocal()            { OrderRepo o; o.Save(); }  // → OrderRepo::Save

    // ── Control: a UNIQUE name resolves with or without the hint ──
    void loadUnique()              { orders_->Load(); }        // → OrderRepo::Load

    // ── Prefer-unknown (§2): must stay unresolved, in BOTH regimes ──
    void chainedReceiver()         { get()->Save(); }          // chained receiver → null

    OrderRepo* get();
};

}  // namespace shop
